### PR TITLE
test: btt tests for BMath

### PR DESCRIPTION
**Pull request by wei3erHase**
Closes BAL-133

### CODE REVIEW

File: test/unit/BMath.t.sol [hunk 1]
```diff
@@ -4,564 +4,263 @@ pragma solidity 0.8.25;
...
 // Main test contract
 contract BMathTest is Test {
   BMath bMath;
-  BNumExposed bNum;
 
-  uint256 BONE; // The fixed-point unit for BMath and BNum
+  uint256 constant BONE = 1e18; // The fixed-point unit for BMath and BNum
+
+  // valid scenario
+  uint256 constant weightIn = BONE;
```
Comment (wei3erHase) [inline test/unit/BMath.t.sol:14]: rm linter rule?

File: test/unit/BMath.t.sol [hunk 2]
```diff
@@ -4,564 +4,263 @@ pragma solidity 0.8.25;
...
-  }
-}
-
 // Main test contract
 contract BMathTest is Test {
   BMath bMath;
-  BNumExposed bNum;
 
-  uint256 BONE; // The fixed-point unit for BMath and BNum
+  uint256 constant BONE = 1e18; // The fixed-point unit for BMath and BNum
```
Comment (0xteddybear) [inline test/unit/BMath.t.sol:11]: why not import from `BConst`?

File: test/unit/BMath.t.sol (hunk 1 above)
Comment (0xteddybear) [reply test/unit/BMath.t.sol:14]: go for it! would you disable it _entirely_ or only for constants? I'd prefer the latter

File: test/unit/BMath.t.sol [hunk 3]
```diff
@@ -4,564 +4,263 @@ pragma solidity 0.8.25;
...
-    );
-
-    // Action
-    uint256 _spotPrice =
-      bMath.calcSpotPrice(_tokenBalanceIn, _tokenWeightIn, _tokenBalanceOut, _tokenWeightOut, _swapFee);
-
-    // Post condition
-    // it should return bi/wi / bo/wo
-    assertEq(_spotPrice, _expected, 'wrong spot price when fee is 0');
+    assertEq(_spotPrice, 1.333333333333333333e18);
```
Comment (0xteddybear) [inline test/unit/BMath.t.sol:87]: I would be okay with a ranged assert as well

File: test/unit/BMath.t.sol [hunk 4]
```diff
@@ -4,564 +4,263 @@ pragma solidity 0.8.25;
...
 
-    // Action
-    uint256 _spotPrice =
-      bMath.calcSpotPrice(_tokenBalanceIn, _tokenWeightIn, _tokenBalanceOut, _tokenWeightOut, _swapFee);
+    uint256 _spotPrice = bMath.calcSpotPrice(balanceIn, weightIn, balanceOut, weightOut, swapFee);
 
-    // Post condition
-    // it should return correct value
-    assertEq(_spotPrice, _expected, 'wrong spot price when fee is not 0');
+    assertEq(_spotPrice, 1.481481481481481481e18);
```
Comment (0xteddybear) [inline test/unit/BMath.t.sol:97]: summing up what we discussed: I think using an easily extendable array of known values and results as a complement of this would be great, but would recommend keeping one testcase outside of it

- as a mitigation on the 'runner' generating false negatives
- a way to see how broken a feature is when tests fail (some edge case in the array failing vs the trivial case)

**Review by 0xteddybear — COMMENTED** [review]
looking good so far, I also beleive this one could be merged to `dev` directly instead of going to a feature branch

File: test/unit/BMath.t.sol (hunk 2 above)
Comment (wei3erHase) [reply test/unit/BMath.t.sol:11]: 🧠

File: test/unit/BMath.t.sol (hunk 3 above)
Comment (wei3erHase) [reply test/unit/BMath.t.sol:87]: i think it doesn't hurt, as it forces the dev to look at the value, copy and paste it and make sure it makes sense, we had some cases in which the tolerance ranges weren't properly set and the test would PASS anyways

**Comment by wei3erHase** [comment]
Closing in favour of #127

File: test/unit/BMath.t.sol (hunk 4 above)
Comment (wei3erHase) [reply test/unit/BMath.t.sol:96]: we can think about later, i believe this abstraction is a step on the right way tho
